### PR TITLE
RustOpaque::Clone has Incorrect Bounds

### LINKED ISSUE #988: Let RustOpaque Expose the Arc
**Is your feature request related to a problem? Please describe.**

Working with async code (via `Stream`s) with this project means that I have to keep objects around even after the function call itself returns. Due to the ownership rules of Rust, I have to use something like `Arc` for this, because both my closure and Dart have to keep a reference to that object around and lifetimes are not transferred over the bridge. `RustOpaque` is already halfway there by using an `Option<Arc<_>>` internally. However, the Rust code doesn't have access to that field.

**Describe the solution you'd like**

`RustOpaque` should implement a function that returns a clone of the inner `Arc`.

**Describe alternatives you've considered**

It's possible to do it right now by using an `Arc` as part of the type wrapped by RustOpaque, but then it's an unnecessary double-`Arc`.

**Additional context**

The PR is incoming, this ticket is just there because you require one for the PR.

### PR DESCRIPTION
**Pull request by anlumo**
Fixes #988

## Checklist

- [X] An issue to be fixed by this PR is listed above.
- [X] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [X] The code generator is run and the code is formatted (e.g. via `just refresh_all`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

Not sure if documentation is needed for this, since it should be self-explanatory.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Given it is draft, I will review after it is marked as ready for review :)

**Comment by fzyzcjy** [comment]
cc @rogurotus - do you think this operation (as_inner) is safe, or will we somehow violate some implicit assumptions?

**Comment by rogurotus** [comment]
> Due to the ownership rules of Rust, I have to use something like Arc for this, because both my closure and Dart have to keep a reference to that object around and lifetimes are not transferred over the bridge.

I didn't quite understand what was going on.
RustOpaque can live entirely on the dart side. (If you pass an object to dart, delete it on the rust side, it will continue to be valid and can be passed back to rust (until dispose is called))

In general, I don’t see any problems, I just don’t quite understand why this is necessary.

**Comment by anlumo** [comment]
> In general, I don’t see any problems, I just don’t quite understand why this is necessary.

Maybe I should supply an example.

Let's say I'm implementing a database in Rust to be accessed from Dart (which is what I'm actually writing right now). All access to the database is asynchronous, because it's using I/O.

`api.rs`:

```rust
pub struct Entry {
    value: String,
}

pub struct Database {
   db: RustOpaque<DatabaseEngine>,
}

impl Database {
    pub fn fetch(&self, sink: StreamSink<Entry>, key: String) {
        let db = self.db.as_inner();
        tokio::spawn(async move {
            let entry = db.get(&key).await.unwrap();
            sink.add(entry);
            sink.close();
        });
    }
}
```

I need to keep `db` around while the closure is running, but I can't do that without having some kind of `Arc` or similar wrapper.

(Note that this obviously isn't my code directly, since there's also error handling and wasm handling and other stuff in there. It's just the structure.)

**Comment by fzyzcjy** [comment]
@anlumo btw will the problem disappear when rust async (https://github.com/fzyzcjy/flutter_rust_bridge/issues/966) is implemented? The rust async looks general and can be widely used

**Comment by anlumo** [comment]
> @anlumo btw will the problem disappear when rust async (#966) is implemented? The rust async looks general and can be widely used

Depends. The Rust borrow checker would be able to handle this case in an async function, but it depends on the lifetime requirements that are required by the executor. For example, if the Future returned from the function would have to be `'static`, it wouldn't work. If the lifetime can be tied to `&self`, it would.

**Comment by rogurotus** [comment]
RustOpaque implements Clone , isn't that what you need?

**Comment by anlumo** [comment]
> RustOpaque implements Clone , isn't that what you need?

I might be able to just use RustOpaque directly, yeah… I have to test that.

**Comment by anlumo** [comment]
> RustOpaque implements Clone , isn't that what you need?

Tried it. It only implements `Clone` when `T` implements `Clone`, so that doesn't work.

The problem is that the code is autoderiving `Clone`, and that's a know issue with the macro that can't be fixed as far as I know.

However, I could fix that within this PR.

**Comment by rogurotus** [comment]
> Tried it. It only implements Clone when T implements Clone, so that doesn't work.

Yes, it looks like a bug. Behavior should be like Arc::clone

**Comment by anlumo** [comment]
> Yes, it looks like a bug. Behavior should be like Arc::clone

I renamed this PR to this instead and also pushed the change to my branch.

**Comment by anlumo** [comment]
I have no idea why my generated files are different. It contains a few new functions where I have no idea what they do.

**Comment by fzyzcjy** [comment]
> I have no idea why my generated files are different. It contains a few new functions where I have no idea what they do.

which ones? the only failure seems only to be a clippy error

image

**Comment by anlumo** [comment]
Never mind, forgot that I added a new test function `run_opaque_inner` which of course generates some more functions.

File: frb_example/pure_dart/rust/src/api.rs
Comment (fzyzcjy) [inline frb_example/pure_dart/rust/src/api.rs:1159]: Well I would suggest just fix that linter error

Comment (anlumo) [reply frb_example/pure_dart/rust/src/api.rs:1159]: The clone is the whole point of that test though.

Comment (fzyzcjy) [reply frb_example/pure_dart/rust/src/api.rs:1159]: Oh I see. Then would be great to add a line of comment to explain it

**Comment by dbsxdbsx** [comment]
> I have no idea why my generated files are different. It contains a few new functions where I have no idea what they do.

Have you tried to run `just precommit` locally before pr?

File: frb_example/pure_dart/rust/src/api.rs
Comment (fzyzcjy) [inline frb_example/pure_dart/rust/src/api.rs:1163]: uber-nit: To make it explicit and avoid future programmers accidentically adding Clone to HideData, what about this: We can create a *new* struct with *explicit* name saying it is not clone. e.g. `struct NonCloneableData { a: String }`, and use it here

**Review by fzyzcjy — COMMENTED** [review]
Good job! I am ready to merge it, if @rogurotus also thinks this PR is OK

File: frb_example/pure_dart/rust/src/api.rs
Comment (anlumo) [reply frb_example/pure_dart/rust/src/api.rs:1163]: Well, I made sure to check that HideData isn't Clone.

Comment (fzyzcjy) [reply frb_example/pure_dart/rust/src/api.rs:1163]: How did you do that? If it is done automatically then also OK.

This is because, we have to imagine that a future programmer may do some refactoring or adding features/tests that may break current code if there is no *automated* tests

Comment (anlumo) [reply frb_example/pure_dart/rust/src/api.rs:1163]: I just did it manually. You're right, if it ever got added, the whole test wouldn't work any more (well, it wouldn't fail, but it'd work with the old code as well).

Comment (fzyzcjy) [reply frb_example/pure_dart/rust/src/api.rs:1163]: Yes that's the problem - then the test will not test what you want to test anymore

**Comment by rogurotus** [comment]
> Good job! I am ready to merge it, if @rogurotus also thinks this PR is OK

Yes, that's how it's supposed to work.

**Comment by fzyzcjy** [comment]
Great!

Then I will merge it after CI is green

**Review by fzyzcjy — APPROVED** [review]

**Comment by fzyzcjy** [comment]
👀

**Comment by anlumo** [comment]
Small update: I got reassigned to a different project for the time being, so I can't continue my work for this.

**Comment by fzyzcjy** [comment]
Take your time! (Btw this one is already OK so after CI is green it can be merged)

**Comment by satvikpendem** [comment]
Any way I can help contribute to this? I was trying to do something similar and would be willing to try fixing the CI. Anything I should do @anlumo and @fzyzcjy?

**Comment by anlumo** [comment]
I'd welcome any help to getting this thing working, but I don't know the best course of action. Should I just give you push access to my fork that this PR is based on? Do you want to have your own fork based on mine and then make a new PR?

**Comment by satvikpendem** [comment]
Push access should be fine, we don't need a new PR since this has relevant details already

**Comment by fzyzcjy** [comment]
Feel free to ping me if need any operations (e.g. make a review) :)

**Comment by satvikpendem** [comment]
Main CI issue looks like the bridge generated files want `use crate::data::NonCloneData;` imported at the top, but they're auto-generated, so I'm not sure where to add this use statement if I'd want the generated files to pick them up. I tried in `api.rs` but it just says that the use statement is unused while still throwing the CI error.

**Comment by fzyzcjy** [comment]
> the bridge generated files want use crate::data::NonCloneData; imported at the top

Not checked the details, but is it possible that, we avoid the need of `use ...` by generating code of *full* names. For example, instead of generating code with name NonCloneData, we generate `flutter_rust_bridge::data::NonCloneData` everywhere

**Comment by satvikpendem** [comment]
Yeah that might solve a lot of issues, with fully qualified names.

**Comment by anlumo** [comment]
@fzyzcjy please merge this PR!

**Comment by fzyzcjy** [comment]
Sure!

**Comment by satvikpendem** [comment]
Awesome, glad to see this happen!

**Comment by fzyzcjy** [comment]
@all-contributors please add @satvikpendem for code

**Comment by satvikpendem** [comment]
Hey @fzyzcjy no need, I didn't actually contribute any code or commits to this PR, it was all @anlumo

**Comment by fzyzcjy** [comment]
@satvikpendem Oh I just checked the discussion history and thought your co-authored. But looks like you submitted at least 1 lines of code? Then you are indeed contributor ;) Since people who only submit 1-char doc fix is contributor as well. At the same time, I document "who did what" in the README, so it is clear what is done by each contributor.

image

**Comment by satvikpendem** [comment]
Haha, alright sounds good then, didn't realize I pushed a commit haha!

**Comment by satvikpendem** [comment]
Ah wait that code seems wrong:

```rust
use crate::data::{
    EnumAlias, Id, MyEnum, MyStruct, NonCloneData, NonCloneData, StructAlias, UserIdAlias,
};
```

The NonCloneData use definition occurs twice, should be a clippy lint warning and removed.

**Comment by anlumo** [comment]
> The NonCloneData use definition occurs twice, should be a clippy lint warning and removed.

I did fix that during my rework yesterday.

**Pull request merged**
